### ISSUE TITLE
CA_RayCastLineAngle()

### ISSUE THREAD

**Issue by Pottus** [Pottus/ColAndreas#4]
This function needs to be updated to return the correct rx/ry/rz angles for in game objects. I've tried a lot of different things but I am still unsure how to accurately translate the normal to an object rotation in that the object would align perpendicular to the orientation of the collision plane.

**Comment by Crayder** [comment]
Kinda lol moment. I'm completely new to the normals and shit, I came up with the following but it doesn't work. I figured it'd be something along those lines. 

	stock SurfaceNormals(Float:P1[3], Float:P2[3], Float:P3[3], &Float:normX, &Float:normY, &Float:normZ, &Float:vectX, &Float:vectY, &Float:vectZ)
	{
		//Normal of three points?
		normX = (P2[0] - P1[0]) * (P3[0] - P1[0]);
		normY = (P2[1] - P1[1]) * (P3[1] - P1[1]);
		normZ = (P2[2] - P1[2]) * (P3[2] - P1[2]);
	
		//Make sum of 'vect' come out as 1, it shouldn't be 0.
		vectX = normX / (floatabs(normX) + floatabs(normY) + floatabs(normZ));
		vectY = normY / (floatabs(normX) + floatabs(normY) + floatabs(normZ));
		vectZ = normZ / (floatabs(normX) + floatabs(normY) + floatabs(normZ));
	}
	
	stock CA_SurfaceNormals(Float:StartX, Float:StartY, Float:StartZ, Float:EndX, Float:EndY, Float:EndZ, &Float:normX, &Float:normY, &Float:normZ, &Float:vectX, &Float:vectY, &Float:vectZ)
	{
		new Float:P1[3], Float:P2[3], Float:P3[3];
		
		//floatsin(120, degrees) = 0.86602540378443864676372317075294
		//floatcos(120, degrees) = -0.5000000000000000000000000000000
		//floattan(120, degrees) = -1.7320508075688772935274463415059
		
		//Get three points on a surface (NOT the best method but yea).
		CA_RayCastLine(StartX, StartY, StartZ, 
			EndX + 0.01, EndY + 0.01, EndZ, 
			P1[0], P1[1], P1[2]);
		CA_RayCastLine(StartX, StartY, StartZ, 
			EndX, EndY + 0.01, EndZ + 0.01, 
			P2[0], P2[1], P2[2]);
		CA_RayCastLine(StartX, StartY, StartZ, 
			EndX + 0.01, EndY, EndZ + 0.01, 
			P3[0], P3[1], P3[2]);
			
		SurfaceNormals(P1, P2, P3, normX, normY, normZ, vectX, vectY, vectZ);
	}
	
The vectors of the CA func always come out to one of the following (it seems):

	0, 0, 0
	0, 1, 0
	0, 0, 1

**Comment by Pottus** [comment]
It is kind of confusing not to say the least.

**Comment by Crayder** [comment]
normX = (P2[0] - P1[0]) * (P3[0] - P1[0]);
normY = (P2[1] - P1[1]) * (P3[1] - P1[1]);
normZ = (P2[2] - P1[2]) * (P3[2] - P1[2]);
Is in short:
N = (P2-P1)*(P3-P1)

Then for the vectors I'm sure I have it incorrect.

**Comment by codectile** [comment]
I don't know what are you all doing, but do you want a calculation for normal of a surface or a mesh. If it is a mesh I have an algorithm for that and, @Crayder did you googled the above algorithm ?

**Comment by codectile** [comment]
@Crayder your algorithm is valid for tri-meshes, if you are ray tracing at three different points to get a  triangle, you are doing it absolutely wrong. What if the object is a sphere or a hull shape ? and your triangle can consist of various other tri-meshes which is totally in accurate.

Getting the normal can be very complicated. For this you will need to get the face where the ray has hit, and then get the vertices of it and then you can calculate the normal using your algorithm. The main complication here, is get the hit face and it's vertices.

**Comment by Crayder** [comment]
1. I said "(NOT the best method but yea)."
2. Yes, that is the best way (yours not mine).

**Comment by codectile** [comment]
Did you google it ? (your method)

**Comment by codectile** [comment]
I think there is a way to do it in bullet.

```
btVector3 normal;
if(rayCall.hasHit()) {
    normal = rayCall.m_hitNormalWorld;
}
```

**Comment by Crayder** [comment]
Did you google it ? (your method)
    
    If you did, that site might contain information about triangles, which you passed out.

Well I googled the small things above; I translated it to x, y, and z.

**Comment by codectile** [comment]
@Pottus Are you looking for something like this: http://vvvv.org/sites/default/files/imagecache/large/images/Normals%20(EX9)%20help-DirectX%20Renderer_2010.12.16-21.19.45.png

If it got it wrong, then please correct me.

**Comment by Crayder** [comment]
Yes, that's what he needs.

**Comment by codectile** [comment]
Okay, so if you want to calculate the angle which is made by the ray and the normal.
You have to do this, I guess.

```c++
if(rayCall.hasHit()) 
{
    btVector3 normal = rayCall.m_hitNormalWorld.angle(vector); //start or end vector of the ray
}
```

**Comment by Pottus** [comment]
Yes picture those lines as objects, I am looking for the correct rotation so that the object would be aligned on the surface perfectly.

I will check you pull requests Saturday thanks, I am away for two days.

**Comment by codectile** [comment]
Okay, now I get you. I am finding a way for it, looks very difficult though.

Yes, do look at the pull request. I have closed two of them because I did some mistake there. Although, there is one open clean request.

**Comment by Crayder** [comment]
What do you mean by "spherical child shape"?

I'm pretty sure every object has a triangular mesh. Look at JernejL's collision mesh layer in Map Construction, it has triangles for every object's collision. These triangle faces are all you would need to do this. Is it possible to get the three points of the face collided in bullet?

**Comment by Crayder** [comment]
For example, these two objects:
![image](https://cloud.githubusercontent.com/assets/9919010/7755597/390de700-ffc3-11e4-9fcb-ddbfd60ead14.png)
Their mesh looks like this:
![image](https://cloud.githubusercontent.com/assets/9919010/7755598/3dc4cc96-ffc3-11e4-96fa-ce71ea265c57.png)
We just need a way to detect the exact face collided and take it's three points.

**Comment by codectile** [comment]
All samp objects consist of meshes but not all gta objects contains meshes, I guess.
Try to check the wireframe of id 2114 I think it is ball.

**Comment by Crayder** [comment]
Object 2114 does have a mesh, without a mesh there would be no shape or anything to apply texture to; but, it doesn't have a collision, and without a collision you can't collide with it in GTA, SA-MP, or CA.

**Comment by codectile** [comment]
Yes, I think renderware consider spherical objects as a polyhedral shape. I am not sure about bullet's consideration. Okay, test it with object, 2887, 19789, 2896, 2942. They don't have meshes at all and they are quite inaccurate in terms of their shape, especially object 2942.

**Comment by Crayder** [comment]
Yes, I think renderware consider spherical objects as a polyhedral shape. I am not sure about bullet's consideration.

The objects have multiple meshes (i.e. collision mesh, and the other one). Also, spheres are not spheres in terms of gta objects.

    Okay, test it with object, 2887, 19789, 2896, 2942. They don't have meshes at all and they are quite inaccurate in terms of their shape, especially object 2942.

Those also have meshes. They are just a few of the exceptional ones that do not have a good collision mesh. SInce their collision mesh is so simplified you could just take three corners and work like a trimesh. You can use any three, they will all give the same vectors.

**Comment by codectile** [comment]
```
The objects have multiple meshes (i.e. collision mesh, and the other one)
```
What's "the other one" ?
Do not confuse yourself with a texture and a mesh.

```
Those also have meshes. They are just a few of the exceptional ones that do not have a good collision mesh.
```
No, they do not have trimeshes. They only consist of textures and collision volumes.
```
SInce their collision mesh is so simplified you could just take three corners and work like a trimesh. You can use any three on the same face, they will all give the same vectors.
```
Well, how will you determine their quad face ?

**Comment by Crayder** [comment]
Well, how will you determine their face ? GTA SA does not provide vertices for collision volumes, just only for meshes.

If that's your argument, then mine is how will you determine the vertices on the trimesh? Apparently it's being done already or we wouldn't have the RayCastLineAngle function.

**Comment by codectile** [comment]
We can determine the vertices of trimesh, they are actualy provided in .col files. I've answered your "argumentative"  question. Now, you answer mine.

For the silos, it clearly shows that they only consist of texture mapping meshes and collision values. If it had collision mesh then it should have a convex hull mesh on the top of it.

**Comment by Crayder** [comment]
We can determine the vertices of trimesh, they are actualy provided in .col files. I've answered your "argumentative" question. Now, you answer mine.
	
I think you just answered it yourself, no need. If those are provided there must be a way to get the vertices of the quad meshes like that the silos have. Like I said before, the silos return perfect angles, and they do not have a trimesh.

	For the silos, it clearly shows that they only consist of texture mapping meshes and collision volumes. If it had collision mesh then it should have a convex hull mesh on the top of it.
	
The outer box is the collision. So aren't we both correct there?

**Comment by codectile** [comment]
We shall not say it as a quad mesh (or poly mesh) because old games contains only trimesh. The silos only have a texture mesh and a collision volume. Collision files does not provide vertices of collision volumes and never will, because we actually do not need it and I did not answer myself, I guess. What angles are you talking about ?

```
The outer box is the collision. So aren't we both correct there?
```
Kinda, you said it's a mesh and I said, it is a collision volume. mesh =/= collision volume

**Comment by Crayder** [comment]
mesh =/= collision volume
	
true

	What angles are you talking about ?

The angles returned by RayCastLineAngle, obviously... :P

**Comment by codectile** [comment]
```
// Instead of returning the position of collision this will return the rx / ry rotation of the slope (I could use use someone to check over the math here)
native CA_RayCastLineAngle(Float:StartX, Float:StartY, Float:StartZ, Float:EndX, Float:EndY, Float:EndZ, &Float:x, &Float:y, &Float:z, &Float:rx, &Float:ry);
```

Yea, it returns the slope so it will be accurate.
According to me slope is: theta = tan^-1(StartY-y/StartX-x)

**Comment by Crayder** [comment]
Wait, so you don't even know what this entire issue thread is about?

The point of this issue is finding a way to get the 'slope' of all surfaces in the collision. That's why we need three points and a normal. Go to the performRayTestAngle function in DynamicWorld.cpp to see where the current math is. It's only X and Y, but it works in most cases (like the silos for example XD). There are a few surfaces it doesn't work for such as the sides of some buildings, fences, some ground, and other objects.

**Comment by codectile** [comment]
Yea, I understood now what he meant by slope.

**Comment by codectile** [comment]
@Pottus is this what you wanted ? Try it and tell me the results.

```c++
btVector3 hit = rayCall.m_hitPointWorld;
btVector3 normal = rayCall.m_hitNormalWorld;
float angle = hit.angle(normal);
float rX = (hit.getX() * cos(angle)) - (hit.getY() * sin(angle));
float rY = (hit.getY() * cos(angle)) + (hit.getX() * sin(angle));
```

**Comment by Crayder** [comment]
http://math.stackexchange.com/questions/1375428/how-can-i-use-the-bullet-physicss-ray-cast-normal-to-calculate-angles-for-a-obj

Not sure if that question should be in the mathematics section, it should probably be in the programming. Whatever though, I kinda claimed the code to get the question through, I mean nothing by that... XD

If anyone reads this, how do you suggest I improve the question?

**Comment by Crayder** [comment]
Basically what we need to do is find the rotation needed to make an object parallel to a surface given the surface's normal.

**Comment by Crayder** [comment]
http://math.etsu.edu/multicalc/prealpha/Chap3/Chap3-6/printversion.pdf

I think this is the math we need. I've been looking all over for this answer! I've asked questions on math.stackexchange, stackoverflow, and even ask. Nobody gives me a good answer, but I found this PDF which is pretty much explaining what we need. 

The tangent of the surface normal.

**Comment by Crayder** [comment]
What we really need is a version of Unity's `Quaternion.FromToRotation`

That Unity function aligns an axis of an object with a surface normal. This is exactly what we need to do. So, does anyone have the source code of that function? :3

**Comment by ikkentim** [comment]
https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/math/Quaternion.java Contains Quaternion functions.

**Comment by Crayder** [comment]
The closest thing I found to `FromToRotation` there was `getAxisAngle`, I don't think it's what we need.

**Comment by Crayder** [comment]
@Pottus 

Perhaps this (4th post) is what we need?

http://www.bulletphysics.org/Bullet/phpBB3/viewtopic.php?f=9&t=10790

If not maybe you could post there and provide a better explanation.

**Comment by Crayder** [comment]
rx = acos(nz);
	ry = 0.0;
	rz = atan2(ny, nx) + 90.0;

I think this is more accurate than the rx/ry version. 

nx, ny, nz: Normal Vector
rx, ry, rz: Return Rotations

Also, the new mesh fix added in 3c5cfc478f06c7d2991d3aa4b6fc2db2b4d5104b seemed to fix the old normal vector bugs (the one that would cause rotations to be off by a few degrees).

**Comment by Pottus** [comment]
I don't think that will work right to be honest currently it works when placing objects on the ground.

**Comment by Crayder** [comment]
It works perfectly, I've tested it a lot and can provide you results.

**Comment by Pottus** [comment]
Lets see

**Comment by Crayder** [comment]
See as in you're going to test it or as in you want to see pictures, or what? XD

**Comment by Pottus** [comment]
Pictures lol

**Comment by Crayder** [comment]
K, a sec

**Comment by Crayder** [comment]
![image](https://cloud.githubusercontent.com/assets/9919010/10777740/ec17fdbe-7cf4-11e5-8c92-7ab0fc60244c.png)

I layed a line of concrete cubes on the hanger, is this good enough? XD

**Comment by Crayder** [comment]
I think I like the rX/rY angle version better still. With the rX/rZ version rotating on the z axis afterwards fucks things up.

**Comment by Pottus** [comment]
Sounds good.

**Comment by codectile** [comment]
Will this be more accurate?

```c++
rx = acos(nx) * RADIAN_TO_DEGREE; //nx normal x of the hit point
ry = asin(ny) * RADIAN_TO_DEGREE; //ny normal y of the hit point
rz = 0.0;
```
